x-pack/plugins/security_solution/server/lib/detection_engine/signals/rule_executor.ts

```diff
@@ -0,0 +1,684 @@
+import { isJobStarted, isMlRule } from '../../../../common/machine_learning/helpers';
+import {
+  isThresholdRule,
+  isEqlRule,
+  isThreatMatchRule,
+} from '../../../../common/detection_engine/utils';
+import { parseScheduleDates } from '../../../../common/detection_engine/parse_schedule_dates';
+import { SetupPlugins } from '../../../plugin';
+import { getInputIndex } from './get_input_output_index';
+import { searchAfterAndBulkCreate } from './search_after_bulk_create';
+import { getFilter } from './get_filter';
+import { RuleAlertAttributes, EqlSignalSearchResponse, WrappedSignalHit } from './types';
+import {
+  getGapBetweenRuns,
+  getListsClient,
+  getExceptions,
+  getGapMaxCatchupRatio,
+  MAX_RULE_GAP_RATIO,
+  wrapSignal,
+  createErrorsFromShard,
+  createSearchAfterReturnType,
+  mergeReturns,
+  createSearchAfterReturnTypeFromResponse,
+  checkPrivileges,
+  hasTimestampFields,
+  hasReadIndexPrivileges,
+} from './utils';
+import { findMlSignals } from './find_ml_signals';
+import { findThresholdSignals } from './find_threshold_signals';
+import { bulkCreateMlSignals } from './bulk_create_ml_signals';
+import { bulkCreateThresholdSignals } from './bulk_create_threshold_signals';
+import { getThresholdBucketFilters } from './threshold_get_bucket_filters';
+import {
+  scheduleNotificationActions,
+  NotificationRuleTypeParams,
+} from '../notifications/schedule_notification_actions';
+import { ruleStatusServiceFactory } from './rule_status_service';
+import { buildRuleMessageFactory } from './rule_messages';
+import { ruleStatusSavedObjectsClientFactory } from './rule_status_saved_objects_client';
+import { getNotificationResultsLink } from '../notifications/utils';
+import { TelemetryEventsSender } from '../../telemetry/sender';
+import { buildEqlSearchRequest } from '../../../../common/detection_engine/get_query_filter';
+import { bulkInsertSignals, filterDuplicateSignals } from './single_bulk_create';
+import { buildSignalFromEvent, buildSignalGroupFromSequence } from './build_bulk_body';
+import { createThreatSignals } from './threat_mapping/create_threat_signals';
+import { getIndexVersion } from '../routes/index/get_index_version';
+import { MIN_EQL_RULE_INDEX_VERSION } from '../routes/index/get_signals_template';
+import { filterEventsAgainstList } from './filters/filter_events_against_list';
+import { isOutdated } from '../migrations/helpers';
+import { RuleTypeParams } from '../types';
+import {
+  AlertExecutorOptions,
+  AlertInstanceContext,
+  AlertInstanceState,
+  AlertTypeState,
+} from '../../../../../alerts/server';
+
+type RuleExecutorParams = AlertExecutorOptions<
+  RuleTypeParams,
+  AlertTypeState,
+  AlertInstanceState,
+  AlertInstanceContext,
+  'default'
+> & {
+  logger: Logger;
+  version: string;
+  ml: SetupPlugins['ml'];
+  lists: SetupPlugins['lists'];
+  eventsTelemetry: TelemetryEventsSender | undefined;
+};
+
+export const ruleExecutor = async ({
```

Comment:
This function seems to be quite long. Does it have a single responsibility? Could refactoring it into smaller pieces improve the readability or maintainability?

Reply:
Just realizing this is a 'move' type PR